### PR TITLE
perf(ts): configuration to potentially speed up nodejs memory usage - PoC

### PR DESCRIPTION
**Pull request by priscilawebdev**
**Type:** Performance 

**Search Terms:** isolatedModules, performance, skipLibCheck

**Description:**

Thanks to my co-worker @Swatinem , we found that activating `skipLibCheck` can considerably speed up the use of nodejs memory.

You can read more about it on his blog:

https://swatinem.de/blog/optimizing-tsc/
https://swatinem.de/blog/js-tooling/

**Before:**

```
Files:          1189
Lines:        484337
Nodes:        892211
Identifiers:  296340
Symbols:      769428
Types:        705065
Memory used: 979716K
I/O read:      0.28s
I/O write:     0.38s
Parse time:    3.46s
Bind time:     1.11s
Check time:   44.64s
Emit time:     3.77s
Total time:   52.99s
```

**After: (skipLibCheck)** 

```
Files:          1189
Lines:        484337
Nodes:        892211
Identifiers:  296340
Symbols:      588373
Types:        636985
Memory used: 902131K
I/O read:      0.58s
I/O write:     0.47s
Parse time:    3.53s
Bind time:     1.06s
Check time:   36.67s
Emit time:     3.76s
Total time:   45.02s
```

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -25,6 +25,9 @@
     "strictBindCallApply": false,
     "experimentalDecorators": true,
 
+    "skipLibCheck": true,
+    "isolatedModules": true,
```
Comment (Swatinem) [inline tsconfig.json]: I don’t think you should use this inside of `tsconfig`. It is more of a setting that you use for `ts-jest` or any kind of bundler plugin. Depending on the project size (und this seems quite big), its a good idea to separate typechecks from other tasks. I will try to blog about this to better explain what I mean.

**Review by Swatinem — DISMISSED** [review]

**Comment by dashed** [comment]
Very nice! 👍

**Comment by dashed** [comment]
Just for context for anyone that's reading over this PR, the optimizations introduced here are, I believe, based on findings from https://swatinem.de/blog/optimizing-tsc/ (written by @Swatinem )

File: tsconfig.json (hunk 1 above)
Comment (billyvg) [reply tsconfig.json]: I agree with this - I think we want to do typechecking of libs during development and we probably want a separate/dedicated CI job that does typechecking while the other jobs do not type check.

**Comment by Swatinem** [comment]
btw, I have further elaborated on my ideas here: https://swatinem.de/blog/js-tooling/ Feedback welcome, it is sometimes hard to articulate my thoughts on all this.

**Review by Swatinem — COMMENTED** [review]
I don’t really know if there are any manually created definition files (`.d.ts`) anywhere?

**Comment by markstory** [comment]
> I don’t really know if there are any manually created definition files (.d.ts) anywhere?

We don't create any, but we use many packages from the definitelytyped project.

**Comment by Swatinem** [comment]
> We don't create any, but we use many packages from the definitelytyped project.

then I guess you should be good to go

**Review by markstory — APPROVED** [review]

**Comment by dashed** [comment]
Looks like this shaved a few more seconds (from 19s to 14s) on top of @BYK 's TS optimizations:

```
$ yarn tsc --noEmit --extendedDiagnostics
yarn run v1.21.1
$ /Users/me/aaa/projects/sentry/sentry/node_modules/.bin/tsc --noEmit --extendedDiagnostics
Files:                       1192
Lines:                     484702
Nodes:                     893383
Identifiers:               296743
Symbols:                   576388
Types:                     144282
Memory used:              555066K
Assignability cache size:   96103
Identity cache size:         6050
Subtype cache size:         18214
I/O Read time:              0.12s
Parse time:                 1.53s
Program time:               2.50s
Bind time:                  0.83s
Check time:                10.62s
Total time:                13.95s
✨  Done in 14.48s.
```

**Comment by BYK** [comment]
Don't want to be the party crasher here but this option reduces type integrity, hence it is not entirely safe to turn on. I see that it yields to an almost 30% improvement in check time but my preference would be towards stricter type checking unless it is prohibitively costly.

This summarizes the issue quite well: https://stackoverflow.com/a/56268670/90297

**Comment by billyvg** [comment]
I agree with @BYK here, aren't we losing type strictness by turning this off?

**Comment by BYK** [comment]
Shall we revert this for the time being? @priscilawebdev @marksteve?

@Swatinem how was your experience regarding this reduced type safety?

**Comment by Swatinem** [comment]
Actually positive, haha. There were cases when external libraries were conflicting over some internals that we didn’t even use. Also, its the job of external libraries to make sure their types are safe, and not yours.

**Comment by priscilawebdev** [comment]
"Most of the time, you can trust the authors of your external dependencies to have done a proper job. Or can you? :-D Anyway, it is very unlikely that you will either mess with, or conflict with any of the definitions that come from node_modules. And setting this flag can considerably speed up your type checking times" by @Swatinem  I'd keep the change

**Comment by dashed** [comment]
> Actually positive, haha. There were cases when external libraries were conflicting over some internals that we didn’t even use. Also, its the job of external libraries to make sure their types are safe, and not yours. ~ @Swatinem 

Out of curiosity, which libraries are these?

-------

I can understand the arguments for both sides. Here's the impact with respect to the Sentry repo.

Types from the repo  https://github.com/DefinitelyTyped/DefinitelyTyped are already rigorously tested against each other across a spectrum of TypeScript versions. This builds a case for enabling `skipLibCheck`.

However, there are libraries which provides their own TS types, and are not tested against the types hosted on the DefinitelyTyped repo. This is where we don't want to enable `skipLibCheck`. However, should a type conflict arise, we would need to address this upstream. 

If we do decide to enable `skipLibCheck` for now, I highly suggest that we run an **optional** frontend build on Travis with `skipLibCheck` disabled. This allows us to have logs on any potential type conflicts between type definitions.

**Comment by BYK** [comment]
>If we do decide to enable skipLibCheck for now, I highly suggest that we run an optional frontend build on Travis with skipLibCheck disabled. This allows us to have logs on any potential type conflicts between type definitions.

@dashed - instead of an additional build, do you think we can add a lint/diff-time rule to check any type libraries outside of DefinitelyTyped?

**Comment by dashed** [comment]
@BYK I'm unsure how to discriminate types like that for checking. 🤔 Do you know if that is possible?

I suggested an extra build to be optional so it won't block any merge.

TS 3.8 is slated to be released in a few weeks which has options for "fast and loose" incremental compilation: https://devblogs.microsoft.com/typescript/announcing-typescript-3-8-rc/#assume-direct-dependencies 

Perhaps we can explore this flag while disabling `skipLibCheck`?

**Comment by Swatinem** [comment]
> Out of curiosity, which libraries are these?

It was back in the days when TS had no versioning support for types, and no `Omit`, where every library defined its own `Omit` type, in the global namespace ofc 🤦‍♂ which then collided with each other.

**Pull request merged**